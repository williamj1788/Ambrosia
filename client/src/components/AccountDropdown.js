import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { clearUser } from '../redux/action'
import s from '../styles/AccountDropdown.module.scss';

class AccountDropdown extends React.Component{
    
    state = {
        redirect: false,
    }
    
    signOutUser = () => {
        fetch('/api/user/signout', {
            credentials: 'include',
        }).then(() => {
            this.props.dispatch(clearUser());
        });
    }

    redirectToOrder = () => {
        this.setState({
            redirect: true,
        });
    }

    componentDidUpdate(){
        if(this.state.redirect){
            this.setState({
                redirect: false
            });
        }
    }
    
    render(){
        if(this.state.redirect){
            return <Redirect to='/user/orders' />
        }
        
        
        let style = {
            height: window.innerWidth > 1000 ? '100px' : '50px'
        };
        if(!this.props.show){
            if(window.innerWidth > 1000){
                style.height = '0';
            }else{
                style.width = '0';
            }
        }

        return(
            <div className={s.dropDown} style={style}>
                <div className={s.content}>
                    <div className={s.tab} onClick={this.redirectToOrder} >
                        <span>Order History</span>
                    </div>
                    <div className={s.tab} onClick={this.signOutUser} >
                        <span>Sign Out</span>
                    </div>
                </div>
            </div>
        )
    }
}

export default connect(state => {
    return {user: state.user}
})(AccountDropdown);