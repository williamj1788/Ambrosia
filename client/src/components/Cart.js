import React from 'react';
import s from '../styles/Cart.module.scss';
import { connect } from 'react-redux';
import { editOrder, removeOrder } from '../redux/action';
import { FaTimes } from 'react-icons/fa';
import uuid from 'uuid';
import { Redirect, withRouter } from 'react-router-dom';

class Cart extends React.Component{
    
    state = {
        active: 1,
        redirect: false,
        error: null,
    }

    handleClick = () => {
        if(!this.props.user){
            this.setRedirect('/signup');
        }else if(this.props.orders.length <= 0){
            this.setState({
                error: 'There are no items'
            });
        }else{
            this.setRedirect('/order');
        }
    }

    removeOrder = id => {
        this.props.dispatch(removeOrder(id));
    }

    editOrder = (id, qty) => {
        this.props.dispatch(editOrder({id,qty}));
    }


    setActive = value => {
        this.setState({active: value});
    }

    setRedirect = value => {
        this.setState({redirect: value});
    }
    
    render(){
        const { redirect, active, error } = this.state;
        const { orders, location, toggle } = this.props;
        if(redirect){
            if(redirect === '/signup' && location.pathname !== '/signup'){
                return <Redirect push to={redirect} />
            }else if(redirect === '/order' && location.pathname !== '/order'){
                return <Redirect push to={redirect} />
            }else{
                toggle();
            }
        }

        return(
            <div className={s.dark}>
                <div className={s.cart}>
                    <Header toggle={toggle} />
                    <OrderContainer 
                    orders={this.props.orders} 
                    remove={this.removeOrder}
                    edit={this.editOrder}
                    active={active}
                    setActive={this.setActive}  />
                    <Footer orders={orders} onClick={this.handleClick} error={error} />
                </div>
            </div>
        )
    }
}

const Header = ({ toggle }) => {
    return(
        <div className={s.header} >
            <span>Cart</span>
            <div className={s.close} onClick={toggle} >
                <FaTimes size='2em' />
            </div>
        </div>
    )
}

const Footer = ({ orders, onClick, error }) => {
    let subTotal = orders.reduce((acc, val) => {
        return acc + (val.price * val.qty);
    }, 0);

    let tax = subTotal * 0.06;

    let deliveryFee = 3;

    let total = subTotal + tax + deliveryFee;

    return(
        <div className={s.footer}>
            <div className={s.flexTotal}>
                <span>Sub Total:</span>
                <span>{'$' + subTotal.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Tax(6%):</span>
                <span>{'$' + tax.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Delivery fees:</span>
                <span>{'$' + deliveryFee.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Total:</span>
                <span>{'$' + total.toFixed(2)}</span>
            </div>
            <button onClick={onClick} type='button' className={s.button} >Checkout</button>
            {error && <p className={s.error}>{error}</p>}
        </div>
    )
}

const OrderContainer = ({ orders, remove, edit, active, setActive }) => {
    const MaxOrders = 7;
    let  curOrders = orders.slice((MaxOrders * (active - 1)), (MaxOrders * active));
    curOrders = curOrders.map((order) => {
        return <Order key={order.id} remove={remove} edit={edit} {...order} />
    });

    if(curOrders.length < 1 && active !== 1){
        setActive(active - 1);
    }

    let counters = [];
    for(let i = 1; i <= Math.ceil(orders.length / MaxOrders); i++){
        counters.push(<button key={uuid()} onClick={() => setActive(i)} className={`${s.counter} ${active === i && s.active}`}>{i}</button>);
    };
    
    return(
        <div className={s.OrderContainer}>
            {curOrders}
            <div className={s.counterCounter}>
                {counters.length > 1 && counters}
            </div>
        </div>
    )
};

const Order = ({ name, price, qty, id, remove, edit }) => {
    return(
        <div className={s.order}>
            <div className={s.orderInfo}>
                <p className={s.orderTitle}>{name}</p>
                <p className={s.orderPrice}>{'$' + price.toFixed(2)}</p>
                <select className={s.select} onChange={event => edit(id, event.target.value)} defaultValue={qty} name="Qty">
                    <option value="1">1</option>
                    <option value="2">2</option>
                    <option value="3">3</option>
                    <option value="4">4</option>
                    <option value="5">5</option>
                    <option value="6">6</option>
                    <option value="7">7</option>
                </select>
            </div>
            <button type='button' className={s.orderRemove} onClick={() => remove(id)} >Remove</button>
        </div>
    )
}

export default withRouter(connect(state => {
    return{
        user: state.user,
        orders: state.orders
    }
})(Cart));