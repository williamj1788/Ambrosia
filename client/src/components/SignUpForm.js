import React from 'react';
import s from '../styles/SignUpForm.module.scss';

class SignUpForm extends React.Component{
    
    state = {
        activeFormBlock: 0,
        formBlockProgress: 0,
        emailError: false
    }

    setActiveFromBlock = target => {
        let newState = {
            activeFormBlock: target
        }
        if(target > this.state.formBlockProgress){
            newState.formBlockProgress = target
        }
        this.setState(newState);
    }

    handleSubmit = () => {
        console.log('TODO');
    }

    validateEmail = () => {
        const { emailError } = this.state;
        const emailValue = document.querySelector('input[name = Email]').value;
        const emailRegex = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
        const isValid = emailRegex.test(emailValue);
        
        if(isValid && emailError){
            this.setState({
                emailError: false,
            });
        }else if(!isValid && !emailError){
            this.setState({
                emailError: 'Invalid Email'
            });
        }
    }
    
    render(){
        const { 
            activeFormBlock,
            formBlockProgress ,
            emailError,
        } = this.state;
        return(
            <div className={s.SignUp}>
                <Header />
                <View>
                    <Form id='signUP-form' active={activeFormBlock} >
                        <FormBlock>
                            <Input type='text' placeholder='Email *' name='Email'onBlur={this.validateEmail} error={emailError} />
                            <Input id='password' type='password' placeholder='Password *' name='Password' />
                            <Input type='password' placeholder='Confirm Password *' name='ConfirmPassword' />
                        </FormBlock>
                        <FormBlock>
                            <Input  type='text' placeholder='First Name *' name='Firstname' />
                            <Input  type='text' placeholder='Last Name *' name='Lastname' />
                            <Input  type='text' placeholder='Address' name='Address' />
                        </FormBlock>
                    </Form>
                </View>
                <Controllers active={activeFormBlock} progress={formBlockProgress} setActive={this.setActiveFromBlock} />
                <ControllerButtons active={activeFormBlock} setActive={this.setActiveFromBlock} submit={this.handleSubmit} />
            </div>
        )
    }
}

const Header = () => {
    return(
        <div className={s.header}>
            <span className={s.headerText}>Sign Up</span>
        </div>
    )
}

const View = ({ children }) => {
    return(
        <div className={s.view}>
            {children}
        </div>
    )
}

const Controllers = ({ active, progress, setActive }) => {
    
    let buttons = [];
    for(let i = 0; i < 2; i++){
        let state;
        if(i === active){
            state = 'active'
        }else if(i <= progress){
            state = 'visted'
        }else{
            state = 'default'
        }
        buttons.push(<button onClick={state === 'visted' ? () => setActive(i) : undefined} className={`${s.controller} ${s[state]}`} ></button>);
    }

    return(
        <div className={s.controllerContainer}>
            {buttons}
        </div>
    )
}

const ControllerButtons = ({ active, setActive, submit }) => {
    return(
        <div className={s.buttonContainer}>
            {active !== 0 && <button className={s.button} onClick={() => setActive(--active)}>Back</button> }
            {active < 1 && <button className={s.button} onClick={() => setActive(++active)}>Next</button>}
            {active === 1 && <button className={s.button} onClick={submit}>Sign Up</button>} 
        </div>
    )
}

const Form = ({ id, active, children }) => {
    const FormWidth = 540; // from SignUpForm.module.scss
    const style = {
        right: FormWidth * active
    };

    
    return(
        <form className={s.form} id={id} style={style}>
            {children}
        </form>
    )
}

const FormBlock = ({ children }) => {
    return(
        <div className={s.formBlock}>
            {children}
        </div>
    )
}

const Input = ({ type, placeholder, onBlur, id, name, error }) => {
    let props = { 
        className: s.input,
        type,
        placeholder,
        name,
        onBlur,
    };
    if(id){
        props.id = id;
    }
    if(error){
        props.style = {
            border: '2px solid red'
        }
    }
    return(
        <div className={s.inputContainer}>
            <input {...props} />
            {error && <p className={s.error}>{error}</p>}
        </div>
    )
}

export default SignUpForm;