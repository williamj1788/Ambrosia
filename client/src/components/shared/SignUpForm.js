import React from 'react';
import s from '../../styles/SignUpForm.module.scss';

import { NavLink, Redirect } from 'react-router-dom';
import GoogleLogin from 'react-google-login';

import { 
    validateEmail,
    validatePassword,
    validateConfirmPassword,
    validateFirstname,
    validateLastname,
 } from '../../validator';

import { connect } from 'react-redux';
import { setUser } from '../../redux/action';

const mapStateToProps = state => {
    return {
        user: state.user
    }
}

export class SignUpForm extends React.Component{
    
    state = {
        activeFormBlock: 0,
        formBlockProgress: 0,
        redirect: false,
        emailError: null,
        passwordError: null,
        confirmPasswordError: null,
        firstnameError: null,
        lastwordError: null,
    }

    handleOnClick = async target => {
        const { activeFormBlock, formBlockProgress } = this.state;
        if(target === activeFormBlock){ // don't unnecessarily set the state
            return
        }
        if(target > formBlockProgress + 1){ // User can only Progress one block at a time 
            return
        }
        if(target > formBlockProgress){
            if(formBlockProgress === 0 && await this.validateFirstFormBlock()){
                this.setState({
                    activeFormBlock: target,
                    formBlockProgress: target
                });
            }
        }else if(target <= formBlockProgress){
            this.setState({
                activeFormBlock: target,
            });
        }
    }

    handleSubmit = async () => {
        const isFirstBlockValid = await this.validateFirstFormBlock();
        if(!isFirstBlockValid){
            this.setState({
                activeFormBlock: 0 
            });
            return
        }
        const isSecondBlockValid = this.validateSecondFormBlock();
        if(!isSecondBlockValid){
            this.setState({
                activeFormBlock: 1 
            });
            return
        }
        await fetch('/api/user/create', {
            credentials: 'include',
            method: 'POST',
            body: this.getFormData(),
        })
        .then(res => res.json())
        .then(res => {
            this.props.dispatch(setUser(res));
        })

        this.setState({
            redirect: true,
        });

        
    }

    validateFirstFormBlock = async () => {
        const isEmailValid = await this.handleEmailBlur();
        const isPasswordValid = this.handlePasswordBlur();
        const isConfirmPassword = this.handleConfirmPasswordBlur();
        
        return isEmailValid && isPasswordValid && isConfirmPassword;
    }

    validateSecondFormBlock = () => {
        const isFirstnameValid = this.handleFirstnameBlur();
        const isLastnameValid = this.handleLastnameBlur();
        
        return isFirstnameValid && isLastnameValid;
    }

    handleEmailBlur = async event => {
        let email;
        if(event){
            email = event.target.value;
        }else{
            email = this.getInputValuebyName('Email');
        }
        const validationError = await validateEmail(email, true);
        
        this.handleValidationErrors(validationError, 'emailError');

        return !validationError;
    }

    handlePasswordBlur = event => {
        let password;
        if(event){
            password = event.target.value;
        }else{
            password = this.getInputValuebyName('Password');
        }
        const validationError = validatePassword(password);

        this.handleValidationErrors(validationError, 'passwordError');
        return !validationError;
    }

    handleConfirmPasswordBlur = event => {
        let confirmPassword;
        if(event){
            confirmPassword = event.target.value;
        }else{
            confirmPassword = this.getInputValuebyName('ConfirmPassword');
        }
        const password = this.getInputValuebyName('Password');
        const validationError = validateConfirmPassword(password, confirmPassword);
        
        this.handleValidationErrors(validationError, 'confirmPasswordError');
        return !validationError;

    }

    handleFirstnameBlur = event => {
        let firstname;
        if(event){
            firstname = event.target.value;
        }else{
            firstname = this.getInputValuebyName('Firstname');
        }
        const validationError = validateFirstname(firstname);

        this.handleValidationErrors(validationError, 'firstnameError');
        return !validationError;
    }

    handleLastnameBlur = event => {
        let lastname;
        if(event){
            lastname = event.target.value;
        }else{
            lastname = this.getInputValuebyName('Lastname');
        }
        const validationError = validateLastname(lastname);

        this.handleValidationErrors(validationError, 'lastwordError');
        return !validationError;
    }

    handleValidationErrors = (validationError, stateErrorProperty) => {
        if(this.state[stateErrorProperty] !== validationError){
            this.setState({
                [stateErrorProperty]: validationError
            });
        }
    }

    onSuccess = res => {
        const token = res.tokenObj.id_token;
        fetch('/api/user/google', {
            method: 'POST',
            headers: {'Authorization': `Bearer ${token}`}
        })
        .then(res => {
            return res;
        })
        .then(res => res.json())
        .then(res => {
            this.props.dispatch(setUser(res));
        })
        .then(() => {
            this.setState({
                redirect: true,
            });
        });
    }
    onFailure = res => {
        console.log(res);
    }
    getInputValuebyName = name => {
        return document.querySelector(`input[name = ${name}]`).value;
    }

    getFormData = () => {
        const form = document.getElementById('signUP-form');
        return new FormData(form);
    }
    
    
    render(){
        const { 
            activeFormBlock,
            formBlockProgress,
            redirect,
            emailError,
            passwordError,
            confirmPasswordError,
            firstnameError,
            lastwordError,
        } = this.state;
        if(redirect || !!this.props.user){
            return <Redirect to='/' />
        }
        return(
            <div className={s.SignUp}>
                <Header title='Sign Up' />
                <View>
                    <Form id='signUP-form' active={activeFormBlock} >
                        <FormBlock>
                            <Input type='text' placeholder='Email *' name='Email' onBlur={this.handleEmailBlur} error={emailError} autocomplete='email' />
                            <Input type='password' placeholder='Password *' name='Password' onBlur={this.handlePasswordBlur} error={passwordError} autocomplete="new-password" />
                            <Input type='password' placeholder='Confirm Password *' name='ConfirmPassword' label="Confirm Password" onBlur={this.handleConfirmPasswordBlur} error={confirmPasswordError} autocomplete="new-password" />
                        </FormBlock>
                        <FormBlock>
                            <Input  type='text' placeholder='First Name *' name='Firstname' onBlur={this.handleFirstnameBlur} error={firstnameError} />
                            <Input  type='text' placeholder='Last Name *' name='Lastname' onBlur={this.handleLastnameBlur} error={lastwordError} />
                            <Input  type='text' placeholder='Address' name='Address' />
                        </FormBlock>
                    </Form>
                </View>
                <Controllers active={activeFormBlock} progress={formBlockProgress} onClick={this.handleOnClick} />
                <ControllerButtons active={activeFormBlock} onClick={this.handleOnClick} submit={this.handleSubmit} />
                <div style={{display: 'flex', margin: '10px 0', justifyContent: 'space-around', height: 50}}>
                    <GoogleLogin
                        clientId="1064409062816-te616f091t5s0vh9mgnkacur1oqrqpr8.apps.googleusercontent.com"
                        onSuccess={this.onSuccess}
                        onFailure={this.onFailure}
                        isSignedIn={true}
                        style={{
                            display: 'block',
                            margin: '0 auto'
                        }}
                    />
                </div>
                <Account />
            </div>
        )
    }
}

export const Header = ({ title }) => {
    return(
        <div className={s.header}>
            <span className={s.headerText}>{title}</span>
        </div>
    )
}

export const View = ({ children }) => {
    return(
        <div className={s.view}>
            {children}
        </div>
    )
}

export const Controllers = ({ active, progress, onClick }) => {
    
    let buttons = [];
    for(let i = 0; i < 2; i++){
        let state;
        if(i === active){
            state = 'active'
        }else if(i <= progress){
            state = 'visted'
        }else{
            state = 'default'
        }
        buttons.push(<button key={i} onClick={state === 'visted' ? () => onClick(i) : undefined} className={`${s.controller} ${s[state]}`} ></button>);
    }

    return(
        <div className={s.controllerContainer}>
            {buttons}
        </div>
    )
}

export const ControllerButtons = ({ active, onClick, submit }) => {
    return(
        <div className={s.buttonContainer}>
            {active !== 0 && <button className={s.button} onClick={() => onClick(active - 1)}>Back</button> }
            {active < 1 && <button className={s.button} onClick={() => onClick(active + 1)}>Next</button>}
            {active === 1 && <button className={s.button} onClick={submit}>Sign Up</button>} 
        </div>
    )
}

export const Form = ({ id, active, children }) => {
    const FormWidth = window.innerWidth < 540 ? window.innerWidth : window.innerWidth > 1980 ? 900 : 540; // from SignUpForm.module.scss
    const style = {
        right: FormWidth * active
    };

    
    return(
        <form className={s.form} id={id} style={style}>
            {children}
        </form>
    )
}

export const FormBlock = ({ children }) => {
    return(
        <div className={s.formBlock}>
            {children}
        </div>
    )
}

export const Input = ({ type, placeholder, label , onBlur, name, error, autocomplete }) => {
    let props = { 
        className: s.input,
        type,
        placeholder,
        name,
        onBlur,
        autoComplete: autocomplete || 'on',
    };
    if(error){
        props.style = {
            border: '2px solid red'
        }
    }
    return(
        <div className={s.inputContainer}>
            <label style={{fontWeight: '500'}} htmlFor={name}>{label || name}</label>
            <input {...props} />
            {error && <p className={s.error}>{error}</p>}
        </div>
    )
}

const Account = () => {
    return(
        <div className={s.account}>
            <p className={s.accountText}>Already have an account?</p>
            <NavLink to='/login'>Login Here</NavLink>
        </div>
    )
}

export default connect(mapStateToProps)(SignUpForm);