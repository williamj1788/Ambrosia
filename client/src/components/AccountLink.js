import React from 'react';
import { connect } from 'react-redux';
import userIcon from '../images/UserIcon.png';
import UserArrow from '../images/Account-arrow.png';
import s from '../styles/Navbar.module.scss';
import { NavLinkWrapper } from './NavComponents';
import AccountDropdown from './AccountDropdown';

export class AccountLink extends React.Component{
    
    state = {
        showDropdown: false,
    }

    componentDidMount(){
        window.addEventListener('scroll', this.handleScroll);
    }

    handleScroll = () => {
        if(this.state.showDropdown){
            this.setState({
                showDropdown: false,
            })
        }
    }
    
    toggleDropdown = () =>{
        this.setState({
            showDropdown: !this.state.showDropdown
        });
    }

    render(){
        const { user } = this.props;
        const { showDropdown } = this.state
        if(user){
            return(
                <div style={{position: 'relative'}} >
                    <div className={s.navLink} style={{cursor: 'pointer'}} onClick={this.toggleDropdown} >
                        <img className={s.linkLogo} src={user.picture ? user.picture : userIcon} alt="User Icon" />
                        <span>{user.firstname}</span>
                        <img className={s.userArrow} src={UserArrow} alt="User Arrow"/>
                    </div>
                    <AccountDropdown show={showDropdown} />
                </div>
            )
        }
        return(
            <NavLinkWrapper to='/login' icon={userIcon}>Login</NavLinkWrapper>
        )
    }
}



export default connect(state => {
    return {user: state.user}
})(AccountLink);