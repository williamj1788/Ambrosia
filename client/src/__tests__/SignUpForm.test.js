import React from 'react';
import { shallow } from 'enzyme';
import { SignUpForm, View, Controllers, ControllerButtons, Form, FormBlock,Input } from '../components/SignUpForm';
import toBeType from "jest-tobetype";
import toJson from 'enzyme-to-json';

expect.extend(toBeType);

describe('<SignUpForm />', () => {
    beforeEach(() => {
        jest.restoreAllMocks();
    });
    const component = shallow(<SignUpForm />);
    const instance = component.instance();
    const event = {
        target:{
            value: 'mock'
        }
    }
    
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });

    test('handleSubmit should set activeFormBlock to zero if firstBlock isnt validated', () => {
        jest.spyOn(instance,'validateFirstFormBlock').mockImplementation(() => false);
        jest.spyOn(instance,'setState')
        instance.handleSubmit();
        expect(instance.setState).toBeCalledWith({activeFormBlock: 0});
    });

    test('handleSubmit should set activeFormBlock to 1 if secondBlock isnt validated', () => {
        jest.spyOn(instance,'validateFirstFormBlock').mockImplementation(() => true);
        jest.spyOn(instance,'validateSecondFormBlock').mockImplementation(() => false);
        jest.spyOn(instance,'setState')
        instance.handleSubmit();
        expect(instance.setState).toBeCalledWith({activeFormBlock: 1});
    });

    test('validateFirstFormBlock should called correct validation functions', () => {
        jest.spyOn(instance,'handleEmailBlur').mockImplementation(() => 'anything');
        jest.spyOn(instance,'handlePasswordBlur').mockImplementation(() => 'anything');
        jest.spyOn(instance,'handleConfirmPasswordBlur').mockImplementation(() => 'anything');
        instance.validateFirstFormBlock();
        expect(instance.handleEmailBlur).toBeCalled();
        expect(instance.handlePasswordBlur).toBeCalled();
        expect(instance.handleConfirmPasswordBlur).toBeCalled();
    });
    test('validateFirstFormBlock should return a boolean', () => {
        jest.spyOn(instance,'handleEmailBlur').mockImplementation(() => true);
        jest.spyOn(instance,'handlePasswordBlur').mockImplementation(() => true);
        jest.spyOn(instance,'handleConfirmPasswordBlur').mockImplementation(() => true);
        instance.validateFirstFormBlock();
        expect(instance.validateFirstFormBlock()).toBeType('boolean');
    });

    test('validateSecondFormBlock should called correct validation functions', () => {
        jest.spyOn(instance,'handleFirstnameBlur').mockImplementation(() => 'anything');
        jest.spyOn(instance,'handleLastnameBlur').mockImplementation(() => 'anything');
        instance.validateSecondFormBlock();
        expect(instance.handleFirstnameBlur).toBeCalled();
        expect(instance.handleLastnameBlur).toBeCalled();
    });
    test('validateSecondFormBlock should return a boolean', () => {
        jest.spyOn(instance,'handleFirstnameBlur').mockImplementation(() => true);
        jest.spyOn(instance,'handleLastnameBlur').mockImplementation(() => true);
        instance.validateSecondFormBlock();
        expect(instance.validateSecondFormBlock()).toBeType('boolean');
    });

    test('handleEmailBlur should call handleValidationErrors with correct args', () => {
        jest.spyOn(instance,'handleValidationErrors');
        jest.spyOn(instance,'validateEmail').mockImplementation(() => 'anything');
        instance.handleEmailBlur(event);
        expect(instance.handleValidationErrors).toHaveBeenCalledWith('anything','emailError');
    });
    test('handleEmailBlur should return a boolean', () => {
        expect(instance.handleEmailBlur(event)).toBeType('boolean');
    });

    test('handlePasswordBlur should call handleValidationErrors with correct args', () => {
        jest.spyOn(instance,'handleValidationErrors');
        jest.spyOn(instance,'validatePassword').mockImplementation(() => 'anything');
        instance.handlePasswordBlur(event);
        expect(instance.handleValidationErrors).toHaveBeenCalledWith('anything','passwordError');
    });
    test('handlePasswordBlur should return a boolean', () => {
        expect(instance.handlePasswordBlur(event)).toBeType('boolean');
    });

    test('handleConfirmPasswordBlur should call handleValidationErrors with correct args', () => {
        jest.spyOn(instance,'handleValidationErrors');
        jest.spyOn(instance,'getInputValuebyName').mockImplementation(() => 'anything');
        jest.spyOn(instance,'validateConfirmPassword').mockImplementation(() => 'anything');
        instance.handleConfirmPasswordBlur(event);
        expect(instance.handleValidationErrors).toHaveBeenCalledWith('anything','confirmPasswordError');
    });
    test('handleConfirmPasswordBlur should return a boolean', () => {
        jest.spyOn(instance,'getInputValuebyName').mockImplementation(() => 'anything');
        expect(instance.handleConfirmPasswordBlur(event)).toBeType('boolean');
    });

    test('handleFirstnameBlur should call handleValidationErrors with correct args', () => {
        jest.spyOn(instance,'handleValidationErrors');
        jest.spyOn(instance,'validateFirstname').mockImplementation(() => 'anything');
        instance.handleFirstnameBlur(event);
        expect(instance.handleValidationErrors).toHaveBeenCalledWith('anything','firstnameError');
    });
    test('handleFirstnameBlur should return a boolean', () => {
        expect(instance.handleFirstnameBlur(event)).toBeType('boolean');
    });

    test('handleLastnameBlur should call handleValidationErrors with correct args', () => {
        jest.spyOn(instance,'handleValidationErrors');
        jest.spyOn(instance,'validateLastname').mockImplementation(() => 'anything');
        instance.handleLastnameBlur(event);
        expect(instance.handleValidationErrors).toHaveBeenCalledWith('anything','lastwordError');
    });
    test('handleLastnameBlur should return a boolean', () => {
        expect(instance.handleLastnameBlur(event)).toBeType('boolean');
    });
    
    test('handleValidationErrors should set the errorState to validationError if they differ', () => {
        component.setState({mockState: null});
        jest.spyOn(instance,'setState');
        instance.handleValidationErrors('mock error', 'mockState');
        expect(instance.setState).toBeCalledWith({'mockState': 'mock error'});
    });

    test("validateEmail should return a string if email has no characters", () => {
        expect(instance.validateEmail('')).toBeType('string');
    });
    test("validateEmail should return a string if email is doesn't regex", () => {
        expect(instance.validateEmail('mock email')).toBeType('string');
    });
    test("validateEmail should return null if email is valid", () => {
        expect(instance.validateEmail('mockEmail123@gmail.com')).toBeNull();
    });

    test("validatePassword should return a string if password has no characters", () => {
        expect(instance.validatePassword('')).toBeType('string');
    });
    test("validatePassword should return a string if password has less than 5 characters", () => {
        expect(instance.validatePassword('mock')).toBeType('string');
    });
    test("validatePassword should return null if password is valid", () => {
        expect(instance.validatePassword('mocky')).toBeNull();
    });

    test("validateConfirmPassword should return a string if password and confirmPassword don't match", () => {
        expect(instance.validateConfirmPassword('mock', 'not mock')).toBeType('string');
    });
    test("validateConfirmPassword should return null if password and confirmPassword match", () => {
        expect(instance.validateConfirmPassword('mock', 'mock')).toBeNull();
    });

    test('validateFirstname should return a string if firstname has no characters', () => {
        expect(instance.validateFirstname('')).toBeType('string');
    });
    test('validateFirstname should return null if is vaild', () => {
        expect(instance.validateFirstname('mock')).toBeNull();
    });

    test('validateLastname should return a string if lastname has no characters', () => {
        expect(instance.validateLastname('')).toBeType('string');
    });
    test('validateLastname should return null if is valid', () => {
        expect(instance.validateLastname('mock')).toBeNull();
    });
});

describe('<View />', () => {
    const component = shallow(<View><div>mock</div></View>);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})

describe('<Controllers />', () => {
    const component = shallow(<Controllers active={0} progress={1} onClick={jest.fn()} />);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})

describe('<ControllerButtons />', () => {
    const component = shallow(<ControllerButtons active={0} submit={jest.fn()} onClick={jest.fn()} />);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})

describe('<Form />', () => {
    const component = shallow(<Form active={1} id={'mockId'}><div>mock children</div></Form>);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})

describe('<FormBlock />', () => {
    const component = shallow(<FormBlock><div>mock</div></FormBlock>);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})

describe('<Input />', () => {
    const component = shallow(<Input type='text' placeholder="mock placeholder" label='mock label' onBlur={jest.fn()} name={'mockName'} error={'mock error'} />);
    test('Snapshot test', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
})