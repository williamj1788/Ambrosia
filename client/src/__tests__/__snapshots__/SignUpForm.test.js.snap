// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControllerButtons /> Snapshot test 1`] = `
<div
  className="buttonContainer"
>
  <button
    className="button"
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;

exports[`<Controllers /> Snapshot test 1`] = `
<div
  className="controllerContainer"
>
  <button
    className="controller active"
    key="0"
  />
  <button
    className="controller visted"
    key="1"
    onClick={[Function]}
  />
</div>
`;

exports[`<Form /> Snapshot test 1`] = `
<form
  className="form"
  id="mockId"
  style={
    Object {
      "right": 540,
    }
  }
>
  <div>
    mock children
  </div>
</form>
`;

exports[`<FormBlock /> Snapshot test 1`] = `
<div
  className="formBlock"
>
  <div>
    mock
  </div>
</div>
`;

exports[`<Input /> Snapshot test 1`] = `
<div
  className="inputContainer"
>
  <label
    htmlFor="mockName"
    style={
      Object {
        "fontWeight": "500",
      }
    }
  >
    mock label
  </label>
  <input
    autoComplete="on"
    className="input"
    name="mockName"
    onBlur={[MockFunction]}
    placeholder="mock placeholder"
    style={
      Object {
        "border": "2px solid red",
      }
    }
    type="text"
  />
  <p
    className="error"
  >
    mock error
  </p>
</div>
`;

exports[`<SignUpForm /> Snapshot test 1`] = `
<div
  className="SignUp"
>
  <Header
    title="Sign Up"
  />
  <View>
    <Form
      active={0}
      id="signUP-form"
    >
      <FormBlock>
        <Input
          autocomplete="email"
          error={null}
          name="Email"
          onBlur={[Function]}
          placeholder="Email *"
          type="text"
        />
        <Input
          autocomplete="new-password"
          error={null}
          name="Password"
          onBlur={[Function]}
          placeholder="Password *"
          type="password"
        />
        <Input
          autocomplete="new-password"
          error={null}
          label="Confirm Password"
          name="ConfirmPassword"
          onBlur={[Function]}
          placeholder="Confirm Password *"
          type="password"
        />
      </FormBlock>
      <FormBlock>
        <Input
          error={null}
          name="Firstname"
          onBlur={[Function]}
          placeholder="First Name *"
          type="text"
        />
        <Input
          error={null}
          name="Lastname"
          onBlur={[Function]}
          placeholder="Last Name *"
          type="text"
        />
        <Input
          name="Address"
          placeholder="Address"
          type="text"
        />
      </FormBlock>
    </Form>
  </View>
  <Controllers
    active={0}
    onClick={[Function]}
    progress={0}
  />
  <ControllerButtons
    active={0}
    onClick={[Function]}
    submit={[Function]}
  />
  <div
    style={
      Object {
        "display": "block",
        "margin": "10px auto",
        "width": "fit-content",
      }
    }
  >
    <e
      accessType="online"
      buttonText="Sign in with Google"
      clientId="1064409062816-te616f091t5s0vh9mgnkacur1oqrqpr8.apps.googleusercontent.com"
      cookiePolicy="single_host_origin"
      disabledStyle={
        Object {
          "opacity": 0.6,
        }
      }
      fetchBasicProfile={true}
      icon={true}
      isSignedIn={true}
      jsSrc="https://apis.google.com/js/api.js"
      onFailure={[Function]}
      onRequest={[Function]}
      onSuccess={[Function]}
      prompt=""
      scope="profile email"
      style={
        Object {
          "display": "block",
          "margin": "0 auto",
        }
      }
      tag="button"
      theme="light"
      type="button"
      uxMode="popup"
    />
  </div>
  <Account />
</div>
`;

exports[`<View /> Snapshot test 1`] = `
<div
  className="view"
>
  <div>
    mock
  </div>
</div>
`;
